### ISSUE TITLE
[torch.compile] Unused differentiable outputs lead to unnecessary gradient computation when compared to eager

### ISSUE THREAD

**Issue by zou3519** [pytorch/pytorch#131794]
The following code:
```py
import torch

@torch.compile(backend="aot_eager")
def f(xs): # E: Function is missing a type annotation  [no-untyped-def]
    result = []
    for x in xs:
        result.append(x.sin())
    return result

xs = [torch.randn([], requires_grad=True) for _ in range(3)]
ys = f(xs)
ys[0].backward()
```
produces the following backward graph:
```
def forward(self, primals_1: "f32[][]cpu", primals_2: "f32[][]cpu", primals_3: "f32[][]cpu", tangents_1: "f32[][]cpu", tangents_2: "f32[][]cpu", tangents_3: "f32[][]cpu"):
     # File: /home/rzou/dev/debug-cpu0/pt-debug-cpu0/foo.py:7 in f, code: result.append(x.sin())
    cos: "f32[][]cpu" = torch.ops.aten.cos.default(primals_3);  primals_3 = None
    mul: "f32[][]cpu" = torch.ops.aten.mul.Tensor(tangents_3, cos);  tangents_3 = cos = None
    cos_1: "f32[][]cpu" = torch.ops.aten.cos.default(primals_2);  primals_2 = None
    mul_1: "f32[][]cpu" = torch.ops.aten.mul.Tensor(tangents_2, cos_1);  tangents_2 = cos_1 = None
    cos_2: "f32[][]cpu" = torch.ops.aten.cos.default(primals_1);  primals_1 = None
    mul_2: "f32[][]cpu" = torch.ops.aten.mul.Tensor(tangents_1, cos_2);  tangents_1 = cos_2 = None
    return [mul_2, mul_1, mul]
```
We're only differentiating through one output, so there should only be a single call to cos. Instead, AOTAutograd is materializing all of the None gradients as zeros.

**Comment by zou3519** [comment]
We should probably emit a warning when this happens and log it.

**Comment by ezyang** [comment]
@bdhirsh this is your backward retracing thing

**Comment by PiesAvalon** [comment]
Here is a small correctness-impacting extension of this issue. The
unused-output behavior can change the gradient of an output that is actually
used, rather than only causing unnecessary backward computation.

```python
import torch


def fn(x):
    return torch.relu(x), torch.sqrt(x)


def get_grad(call):
    x = torch.tensor([-1.0, 1.0], requires_grad=True)
    used, unused = call(x)
    return torch.autograd.grad(used.sum(), x)[0]


print("native eager:", get_grad(fn))

for backend in ["eager", "aot_eager", "inductor"]:
    compiled = torch.compile(fn, backend=backend, fullgraph=True)
    print(f"{backend:>10}:", get_grad(compiled))
```

On the current nightly,
`2.15.0.dev20260820+cu130`
(`a69ffdbc0fa11bf7dc90221fba0abe4b106fa214`), this prints:

```text
native eager: tensor([0., 1.])
     eager: tensor([0., 1.])
 aot_eager: tensor([nan, 1.])
  inductor: tensor([nan, 1.])
```

The `sqrt` result is returned by `fn` but is never used in the loss. Native
eager autograd therefore does not traverse that output's backward branch and
returns the expected ReLU gradient. With AOTAutograd, the unused output is
given a materialized zero cotangent and its backward computation is still
executed. At `x = -1`, the non-finite `sqrt` derivative combines with that zero
cotangent and introduces a spurious NaN into the gradient of the used ReLU
output. Inductor shows the same result because its backward passes through
AOTAutograd.

This is deterministic on CPU and does not involve randomness or numerical
tolerances. It appears to have the same root cause as this issue, but
demonstrates a silent correctness consequence: enabling AOTAutograd changes a
finite used gradient into NaN.

The general `0 * NaN` autograd behavior is also discussed in #4132, but the
eager/AOTAutograd discrepancy here is specifically caused by traversing an
entire output branch that eager autograd leaves unused.
